### DISCUSSION TITLE
[Proposal] Use tuple special syntax in generics constraints

### DISCUSSION THREAD

**Discussion by MadL1me** [dotnet/csharplang#5774] [Language Ideas] [answered]
# Use tuple special syntax in generics constraints

- [x] Proposed
- [ ] Prototype: Not Started
- [ ] Implementation: Not Started
- [ ] Specification: Not Started

# Summary
Use tuple special syntax in generics constraints

# Motivation
Stating from C# 7.0 we get used to special tuple syntax:
```cs 
(int,int) tuple = (1, 2);  // can create

public (int, float) GetTwoNumbers()  // can get from method
{
    return (1, 5.5);
}

public void DoStuff((int,int) param)  // can pass to method
{
    //do stuff
}
```

However, I've noticed some inconsistency, because I can't use same thing in generics type constraint: 

```cs
public void GenericMethod<T>() where T: (int,int) //Error: only a class or interface can be specified as constraint
{
    //do stuff
}
```

Yes, we can define tuple constraint this way, but its still "visually unefficient", at least for me: 

```cs
public void GenericMethod<T>() where T: Tuple<int,int> 
{
    //do stuff
}
```

**Comment by CyrusNajmabadi** [comment]
What would be the point of this constraint?  As is a value type, there can't be any inheritance here.  It would be like doing `where T : DateTime` (also not legal).

**Reply by jnm2** [reply]
To put it another way, why would you want to do that, when you can do this?

```cs
public void GenericMethod()
{
    // Use `typeof(ValueTuple<int, int>)` instead of `typeof(T)`
}
```

**Comment by Richiban** [comment, marked as answer]
`(int, int)` and `Tuple<int, int>` are actually not the same; `(int, int)` means `ValueTuple<int, int>` which is a struct and therefore cannot be inherited from.

Having said that, I'm not sure I understand why you would want this. You've given 

```cs
public void GenericMethod<T>() where T: Tuple<int,int> 
{
    //do stuff
}
```

as an example; do you write types that inherit from `Tuple<>`? Have you noticed that you can inherit from `Tuple<int, int>` but not `(int, int)`?

**Reply by MadL1me** [reply]
Oh, I didn't know this, thanks you for an answer
